The agent panics when UDP port is busy and dogstatsd_so_rcvbuf are configured

There is a bug [here](https://github.com/DataDog/datadog-agent/blob/fbe415d9ec9875e173525b7234911e502ab33bf0/pkg/dogstatsd/listeners/udp.go#L65-L75): the connection returned from `net.ListenUDP("udp", addr)` is used before the error check is performed. Here is what that results in:
```
2020-11-19 02:33:54 UTC | CORE | INFO | (cmd/agent/app/run.go:360 in StopAgent) | See ya!
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2c8925c]

goroutine 1 [running]:
github.com/DataDog/datadog-agent/pkg/dogstatsd/listeners.NewUDPListener(0xc00098f380, 0xc0007d3e60, 0xe, 0x1fbd, 0x0)
	/.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/dogstatsd/listeners/udp.go:68 +0x5fc
github.com/DataDog/datadog-agent/pkg/dogstatsd.NewServer(0xc0008485a0, 0x3a1af38, 0xd, 0x1)
	/.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/dogstatsd/server.go:158 +0x1100
github.com/DataDog/datadog-agent/cmd/agent/app.StartAgent(0xc000000008, 0x3b17538)
	/.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/cmd/agent/app/run.go:275 +0x126e
github.com/DataDog/datadog-agent/cmd/agent/app.run(0x5627f00, 0x580e2f0, 0x0, 0x0, 0x0, 0x0)
	/.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/cmd/agent/app/run.go:122 +0x1de
github.com/spf13/cobra.(*Command).execute(0x5627f00, 0x580e2f0, 0x0, 0x0, 0x5627f00, 0x580e2f0)
	/.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/vendor/github.com/spf13/cobra/command.go:842 +0x460
github.com/spf13/cobra.(*Command).ExecuteC(0x56232e0, 0x0, 0x4f180, 0xc0000e2058)
	/.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/vendor/github.com/spf13/cobra/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
	/.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/vendor/github.com/spf13/cobra/command.go:887
main.main()
	/.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/cmd/agent/main.go:18 +0x2d
```

--- a/pkg/dogstatsd/listeners/udp.go
+++ b/pkg/dogstatsd/listeners/udp.go
@@ -63,15 +63,14 @@ func NewUDPListener(packetOut chan Packets, sharedPacketPool *PacketPool) (*UDPL
 		return nil, fmt.Errorf("could not resolve udp addr: %s", err)
 	}
 	conn, err := net.ListenUDP("udp", addr)
+	if err != nil {
+		return nil, fmt.Errorf("can't listen: %s", err)
+	}
 
 	if rcvbuf := config.Datadog.GetInt("dogstatsd_so_rcvbuf"); rcvbuf != 0 {
 		if err := conn.SetReadBuffer(rcvbuf); err != nil {
 			return nil, fmt.Errorf("could not set socket rcvbuf: %s", err)
 		}
-	}
-
-	if err != nil {
-		return nil, fmt.Errorf("can't listen: %s", err)
 	}
 
 	bufferSize := config.Datadog.GetInt("dogstatsd_buffer_size")

--- a/pkg/dogstatsd/listeners/udp_test.go
+++ b/pkg/dogstatsd/listeners/udp_test.go
@@ -141,6 +141,23 @@ func TestUDPReceive(t *testing.T) {
 	}
 }
 
+// Reproducer for https://github.com/DataDog/datadog-agent/issues/6803
+func TestNewUDPListenerWhenBusyWithSoRcvBufSet(t *testing.T) {
+	port, err := getAvailableUDPPort()
+	assert.Nil(t, err)
+	address, _ := net.ResolveUDPAddr("udp", fmt.Sprintf("127.0.0.1:%d", port))
+	conn, err := net.ListenUDP("udp", address)
+	assert.NotNil(t, conn)
+	assert.Nil(t, err)
+	defer conn.Close()
+	config.Datadog.SetDefault("dogstatsd_so_rcvbuf", 1)
+	config.Datadog.SetDefault("dogstatsd_port", port)
+	config.Datadog.SetDefault("dogstatsd_non_local_traffic", false)
+	s, err := NewUDPListener(nil, packetPoolUDP)
+	assert.Nil(t, s)
+	assert.NotNil(t, err)
+}
+
 // getAvailableUDPPort requests a random port number and makes sure it is available
 func getAvailableUDPPort() (int, error) {
 	conn, err := net.ListenPacket("udp", ":0")

--- a/releasenotes/notes/dogstatsd-Fix-error-handling-in-ListenUDP-288a788984dd1750.yaml
+++ b/releasenotes/notes/dogstatsd-Fix-error-handling-in-ListenUDP-288a788984dd1750.yaml
@@ -1,0 +1,11 @@
+# Each section from every release note are combined when the
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+fixes:
+  - |
+    Fix agent panic when UDP port is busy and dogstatsd_so_rcvbuf is configured.

